### PR TITLE
[quantization] Implement QuantGemma4ForCausalLM PTQ wrapper

### PR DESCRIPTION
**Pull request by dvsav**
## What

This PR implements the full PTQ calibration support for `Gemma4ForCausalLM` — the text-only causal LM that wraps `Gemma4TextModel` (text decoder) and adds an `lm_head` linear layer with logit softcapping. Previously this wrapper was a skeleton stub; now it has a complete calibration path with logit softcapping and fake-quantization observers, mirroring the implementation already done for `QuantGemma4ForConditionalGeneration`.

## Why

The `QuantGemma4ForCausalLM` wrapper existed only as a skeleton — its `forward()` returned raw logits without logit softcapping, and the module was commented out in the registry. This meant the text-only Gemma4 causal LM could not be quantized. This PR completes the wrapper's calibration path so the prepare → calibrate → convert pipeline works, matching the behavior of the original `Gemma4ForCausalLM.forward` including logit softcapping.

Per the [Support Gemma4](https://github.com/Samsung/TICO/issues/768) issue specification, the full Hugging Face Gemma4 forward is **not** exported as one NPU graph. Instead, the Llama static runtime pattern is followed, where `StaticGemma4Runtime` orchestrates the export of individual submodules. Therefore, this wrapper intentionally does **not** provide `as_export_module()` or `forward_export()` methods — the static runtime handles export at the submodule level.

## Key Design Decisions

1. **No export adapter** — This wrapper does not provide `as_export_module()` or `forward_export()`. The static runtime (`StaticGemma4Runtime`) exports individual submodules (text decoder layers, etc.) rather than the full model. The old skeleton `as_export_module()` that returned `Gemma4LMHeadExportAdapter` was removed.

2. **Three observers for the softcapping path** — `obs_logit_softcapping_div` (after `logits / softcap`), `obs_logit_softcapping_tanh` (after `tanh`), and `obs_logits` (final logits). Each intermediate tensor in the softcapping chain gets its own observer so that every graph node carries qparam metadata.

3. **Extracted softcapping logic** — The div → fq → tanh → fq → mul → fq chain is encapsulated in a single `_apply_logit_softcapping()` helper method, keeping the `forward()` method clean and ensuring the softcapping logic is defined in one place.

4. **`logits_to_keep` tensor handling** — `forward()` matches the original's behavior: `int` → slice, `torch.Tensor` → advanced indexing, instead of silently falling through to `slice(None)` for tensor inputs.

5. **`config.final_logit_softcapping` vs `config.get_text_config()`** — Unlike `QuantGemma4ForConditionalGeneration` (which uses `self.config.get_text_config().final_logit_softcapping` because its config is a top-level `Gemma4Config`), this wrapper accesses `self.config.final_logit_softcapping` directly because `Gemma4ForCausalLM`'s config is already a `Gemma4TextConfig`.

6. **`assert_gemma4_e2b_no_moe` guard** — Added the same E2B/no-MoE assertion used by `QuantGemma4ForConditionalGeneration` to fail fast on unsupported model variants.

7. **Calibration-only return type** — The wrapper returns raw logits (not `Gemma4CausalLMOutputWithPast`). This is sufficient for PTQ calibration where only logits matter for observer statistics. A TODO documents this for future HF compatibility.

## Changes

- **`tico/quantization/wrapq/wrappers/gemma4/quant_for_causal_lm.py`** — Implemented `forward()` with logit softcapping and `logits_to_keep` tensor handling; added `_apply_logit_softcapping()` helper with three fake-quantization observers; added `generate()` delegation; added `assert_gemma4_e2b_no_moe` guard in `__init__`; updated `_all_observers()` to return the three new observers; removed the old skeleton `as_export_module()`.
- **`tico/quantization/wrapq/wrappers/registry.py`** — Enabled the `quant_for_causal_lm` module in `_CORE_MODULES` (uncommented).
- **`test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py`** — New file: unit tests covering wrapper type, forward finiteness, output shape, `logits_to_keep`, softcapping bounds, prepare-convert flow, and observer calibration.
- **`test/quantization/wrapq/wrappers/gemma4/test_quantize_for_causal_lm.py`** — New file: smoke tests for the prepare-calibrate-convert flow, including FP parity, `logits_to_keep`, and softcapping bounds.
- **`tico/quantization/wrapq/examples/gemma4/quantize_for_causal_lm.py`** — New file: example script demonstrating the PTQ flow (prepare → calibrate → convert → PEIR comparison).
- **`tico/quantization/recipes/debug/wrapper_smoke/cases/gemma4.py`** — Added `Gemma4ForCausalLMCase` smoke case and registered it in `GEMMA4_CASES`.

## Tests

All tests were run with `RUN_INTERNAL_TESTS=1`:
- `test_quant_for_causal_lm.py` — Validates wrapper type, forward finiteness, output shape `(batch, seq_len, vocab_size)`, `logits_to_keep=1` slicing, softcapping bounds (`|logits| ≤ softcap`), full prepare-convert flow, and observer calibration after convert.
- `test_quantize_for_causal_lm.py` — Validates FP parity before quantization, prepare-convert flow with text-only calibration data, `logits_to_keep` slicing, and softcapping bounds.
- Wrapper smoke runner — Validates the `gemma4_for_causal_lm` smoke case passes with acceptable quantization error.

### Unit Tests

```
$ RUN_INTERNAL_TESTS=1 python -m pytest test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py -v
========================================================== test session starts ===========================================================
platform linux -- Python 3.10.12, pytest-9.0.3, pluggy-1.6.0 -- /home/d.savchenkov/myenv/bin/python
cachedir: .pytest_cache
rootdir: /home/d.savchenkov/TICO
configfile: pyproject.toml
plugins: anyio-4.13.0
collected 7 items                                                                                                                        

test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py::TestQuantGemma4ForCausalLMSmoke::test_forward_output_shape PASSED [ 14%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py::TestQuantGemma4ForCausalLMSmoke::test_logit_softcapping_applied PASSED [ 28%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py::TestQuantGemma4ForCausalLMSmoke::test_logits_to_keep PASSED   [ 42%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py::TestQuantGemma4ForCausalLMSmoke::test_observers_calibrated PASSED [ 57%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py::TestQuantGemma4ForCausalLMSmoke::test_prepare_convert_flow PASSED [ 71%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py::TestQuantGemma4ForCausalLMSmoke::test_prepare_returns_correct_wrapper PASSED [ 85%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_causal_lm.py::TestQuantGemma4ForCausalLMSmoke::test_text_only_forward_is_finite PASSED [100%]

=========================================================== 7 passed in 50.03s ===========================================================
```

### Internal Tests

```
$ RUN_INTERNAL_TESTS=1 python -m pytest test/quantization/wrapq/wrappers/gemma4/test_quantize_for_causal_lm.py -v
========================================================== test session starts ===========================================================
platform linux -- Python 3.10.12, pytest-9.0.3, pluggy-1.6.0 -- /home/d.savchenkov/myenv/bin/python
cachedir: .pytest_cache
rootdir: /home/d.savchenkov/TICO
configfile: pyproject.toml
plugins: anyio-4.13.0
collected 4 items                                                                                                                        

test/quantization/wrapq/wrappers/gemma4/test_quantize_for_causal_lm.py::TestGemma4ForCausalLMSmoke::test_logit_softcapping_bounded PASSED [ 25%]
test/quantization/wrapq/wrappers/gemma4/test_quantize_for_causal_lm.py::TestGemma4ForCausalLMSmoke::test_logits_to_keep PASSED     [ 50%]
test/quantization/wrapq/wrappers/gemma4/test_quantize_for_causal_lm.py::TestGemma4ForCausalLMSmoke::test_no_quant_model_matches_reference PASSED [ 75%]
test/quantization/wrapq/wrappers/gemma4/test_quantize_for_causal_lm.py::TestGemma4ForCausalLMSmoke::test_prepare_convert_flow PASSED [100%]

=========================================================== 4 passed in 30.62s ===========================================================
```

### Smoke Test

```
$ python -m tico.quantization.examples.inspect \
    --config tico/quantization/examples/configs/wrapper_smoke.yaml \
    --mode wrapper-smoke \
    --case gemma4_for_causal_lm \
    --output-dir ./out/wrapper_smoke
┌───────────── Wrapper Smoke Summary ─────────────
│ Case             : gemma4_for_causal_lm
│ Status           : PASS
│ Mean |diff|      : 0.000540
│ Max |diff|       : 0.024903
│ PEIR             : 0.021237
│ Shape match      : True
│ Quant finite     : True
└─────────────────────────────────────────────────
     ┌───────────────────────────────────────────┐
 0.62┤                                           │
     │                                       ••  │
     │                                    •••    │
 0.40┤                                 ••••      │
     │                               •••         │
     │                            ••••           │
 0.19┤                          ••••             │
     │                       ••••                │
-0.03┤                     •••                   │
     │                  ••••                     │
     │                ••••                       │
-0.24┤              •••                          │
     │           ••••                            │
     │         •••                               │
-0.46┤      ••••                                 │
     │     ••                                    │
     │  •                                        │
-0.67┤                                           │
     └┬──────────┬─────────┬──────────┬─────────┬┘
    -0.67      -0.35     -0.03      0.30     0.62 
```

## Example Script

**`tico/quantization/wrapq/examples/gemma4/quantize_for_causal_lm.py`** — Demonstrates the PTQ workflow:
1. Creates a tiny `Gemma4ForCausalLM` with `final_logit_softcapping=30.0` (no download needed).
2. Prepares the model with `build_gemma4_e2b_ptq_config` (with `num_vision_layers=0` since this is text-only).
3. Calibrates with 20 synthetic text-only samples.
4. Converts to a fake-quantized model.
5. Compares FP vs. quantized logits (prints PEIR and a plot).

Note: This script does not export to Circle format, as the full model export is handled by `StaticGemma4Runtime` at the submodule level, not by this wrapper.

```
$ python tico/quantization/wrapq/examples/gemma4/quantize_for_causal_lm.py
Preparing model for quantization...
Calibrating (text-only)...
Converting to quantized model...

┌───────────── Quantization Error Summary ─────────────
│ FP output shape    : (1, 16, 256)
│ Quant output shape : (1, 16, 256)
│ Mean |diff|        : 0.000070
│ PEIR               : 0.052900 %
└──────────────────────────────────────────────────────
     ┌───────────────────────────────────────────┐
 0.60┤                                           │
     │                                      •••  │
 0.39┤                                   •••     │
     │                                ••••       │
     │                             ••••          │
 0.19┤                           •••             │
     │                        ••••               │
-0.02┤                     ••••                  │
     │                  ••••                     │
     │               ••••                        │
-0.23┤             •••                           │
     │          ••••                             │
-0.44┤       ••••                                │
     │    •••                                    │
     │  •                                        │
-0.65┤                                           │
     └┬──────────┬─────────┬──────────┬─────────┬┘
    -0.65      -0.34     -0.02      0.29     0.60
```

### CODE REVIEW

**Review by Torrero — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]
LGTM
